walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,67 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<Event> sortedEvents = new ArrayList<Event>(events);
+    Collection<TimeRange> freeTimeRanges = new ArrayList<TimeRange>();
+    Collection<TimeRange> freeTimeRangesWithOptionalAttendee = new ArrayList<TimeRange>();
+    int endOptional = 0, endMandatory = 0, FULL_DAY = 24*60;
+    
+    Collections.sort(sortedEvents, new Comparator<Event>() {
+      @Override public int compare(Event e1, Event e2) {
+        return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+      }
+    });
+    
+    for(Event e: sortedEvents) {
+      if(!Collections.disjoint(e.getAttendees(), request.getAttendees())) {
+        if(isOverlapping(e, endMandatory, request.getDuration())) {
+          freeTimeRanges.add(TimeRange.fromStartEnd(endMandatory, e.getWhen().start(), false));
+        }
+
+        if(endMandatory < e.getWhen().end()) {
+          endMandatory = e.getWhen().end();
+        }
+      }
+
+      if(!Collections.disjoint(e.getAttendees(), request.getAttendees()) || !Collections.disjoint(e.getAttendees(), request.getOptionalAttendees())) {
+        if(isOverlapping(e, endOptional, request.getDuration())) {
+          freeTimeRangesWithOptionalAttendee.add(TimeRange.fromStartEnd(endOptional, e.getWhen().start(), false));
+        }
+
+        if(endOptional < e.getWhen().end()) {
+          endOptional = e.getWhen().end();
+        }
+      }
+
+      
+    }
+
+    //If last event doesn't go to the end of the day
+    if(endMandatory < FULL_DAY && request.getDuration() < FULL_DAY && !request.getAttendees().isEmpty()) {
```

Comment:
I believe there is constant like END_OF_DAY already defined. Can you use that?

Reply:
Unfortunately it is defined to the value of 1 less than what I need